be/src/util/cgroup_util.cpp

```diff
@@ -0,0 +1,229 @@
+
+using strings::CUnescape;
+using strings::Split;
+using strings::SkipWhitespace;
+using std::pair;
+
+namespace doris {
+
+Status CGroupUtil::find_global_cgroup(const string& subsystem, string* path) {
+    ifstream proc_cgroups("/proc/self/cgroup", ios::in);
+    string line;
+    while (true) {
+        if (proc_cgroups.fail()) {
+            return Status::IOError(Substitute("Error reading /proc/self/cgroup: $0", get_str_err_msg()));
+        } else if (proc_cgroups.peek() == std::ifstream::traits_type::eof()) {
+            return Status::NotFound(Substitute("Could not find subsystem $0 in /proc/self/cgroup", subsystem));
+        }
+        // The line format looks like this:
+        // 4:memory:/user.slice
+        // 9:cpu,cpuacct:/user.slice
+        // so field size will be 3
+        getline(proc_cgroups, line);
+        if (!proc_cgroups.good()) {
+            continue;
+        }
+        vector<string> fields = Split(line, ":");
+        DCHECK_GE(fields.size(), 3);
+        // ":" in the path does not appear to be escaped - bail in the unusual case that
+        // we get too many tokens.
+        if (fields.size() != 3) {
+            return Status::InvalidArgument(Substitute(
+                                               "Could not parse line from /proc/self/cgroup - had $0 > 3 tokens: '$1'",
+                                               fields.size(), line));
+        }
+        vector<string> subsystems = Split(fields[1], ",");
+        auto it = std::find(subsystems.begin(), subsystems.end(), subsystem);
+        if (it != subsystems.end()) {
+            *path = move(fields[2]);
+            return Status::OK();
+        }
+    }
+}
+
+static Status unescape_path(const string& escaped, string* unescaped) {
+    string err;
+    if (!CUnescape(escaped, unescaped, &err)) {
+        return Status::InvalidArgument(Substitute("Could not unescape path '$0': $1", escaped, err));
+    }
+    return Status::OK();
+}
+
+static Status read_cgroup_value(const string& limit_file_path, int64_t* val) {
+    ifstream limit_file(limit_file_path, ios::in);
+    string line;
+    getline(limit_file, line);
+    if (limit_file.fail() || limit_file.bad()) {
+        return Status::IOError(Substitute("Error reading $0: $1", limit_file_path, get_str_err_msg()));
+    }
+    StringParser::ParseResult pr;
+    // Parse into an an int64_t If it overflows, returning the max value of int64_t is ok because that
+    // is effectively unlimited.
+    *val = StringParser::string_to_int<int64_t>(line.c_str(), line.size(), &pr);
+    if ((pr != StringParser::PARSE_SUCCESS && pr != StringParser::PARSE_OVERFLOW)) {
+        return Status::InvalidArgument(
+                   Substitute("Failed to parse $0 as int64: '$1'", limit_file_path, line));
+    }
+    return Status::OK();
+}
+
+Status CGroupUtil::find_cgroup_mounts(
+    const string& subsystem, pair<string, string>* result) {
+    ifstream mountinfo("/proc/self/mountinfo", ios::in);
```

Comment:
Is there any resouce leak if ifstream.close is not called?

Reply:
this is done automatically by the ifstream destructor.